rest/config_manager.go

```diff
@@ -0,0 +1,639 @@
+// the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/db"
+)
+
+// ConfigManager should be used for any read/write of persisted database configuration files
+type ConfigManager interface {
+	// GetConfig fetches a database config for a given bucket and config group ID, along with the CAS of the config document. Does not enforce version match with registry.
+	GetConfig(bucket, groupID, dbName string, config *DatabaseConfig) (cas uint64, err error)
+	// GetDatabaseConfigs returns all configs for the bucket and config group.  Enforces version match with registry.
+	GetDatabaseConfigs(ctx context.Context, bucketName, groupID string) ([]*DatabaseConfig, error)
+	// InsertConfig saves a new database config for a given bucket and config group ID.
+	InsertConfig(ctx context.Context, bucket, groupID string, config *DatabaseConfig) (newCAS uint64, err error)
+	// UpdateConfig updates an existing database config for a given bucket and config group ID. updateCallback can return nil to remove the config.
+	UpdateConfig(ctx context.Context, bucket, groupID, dbName string, updateCallback func(bucketConfig *DatabaseConfig) (updatedConfig *DatabaseConfig, err error)) (newCAS uint64, err error)
+	// DeleteConfig removes a database config for a given bucket, config group ID and database name.
+	DeleteConfig(ctx context.Context, bucket, dbName, groupID string) (err error)
+}
+
+type dbConfigNameOnly struct {
+	Name string `json:"name"`
+}
+
+var _ ConfigManager = &bootstrapContext{}
+
+const configUpdateMaxRetryAttempts = 5 // Maximum number of retries due to conflicting updates or rollback
+const configFetchMaxRetryAttempts = 5  // Maximum number of retries due to registry rollback
+
+// GetConfig fetches a database name for a given bucket and config group ID.
+func (b *bootstrapContext) GetConfigName(bucketName, groupID, dbName string, configName *dbConfigNameOnly) (cas uint64, err error) {
+	return b.Connection.GetMetadataDocument(bucketName, PersistentConfigKey(groupID, dbName), configName)
+}
+
+// GetConfig fetches a database config for a given bucket and config group ID, along with the CAS of the config document.
+// GetConfig does *not* validate that config version matches registry version - operations requiring synchronization
+// with registry should use getRegistryAndDatabase, or getConfig with the required version
+func (b *bootstrapContext) GetConfig(bucketName, groupID, dbName string, config *DatabaseConfig) (cas uint64, err error) {
+	return b.Connection.GetMetadataDocument(bucketName, PersistentConfigKey(groupID, dbName), config)
+}
+
+// InsertConfig saves a new database config for a given bucket and config group ID. This is a three-step process:
+//  1. getRegistryAndDatabase to enforce synchronization pre-update
+//  2. Update the registry to add the config
+//  3. Write the config
```

Comment:
Given the multi-step/complexity in these functions I'm wondering whether it's worth logging at each step for future diagnosis, or at least in the outer retry loop.

Reply:
Agree - added this logging.